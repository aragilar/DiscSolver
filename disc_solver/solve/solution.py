# -*- coding: utf-8 -*-
"""
The system of odes
"""

from math import pi, sqrt, tan, degrees, radians

import logbook

import numpy as np

from scikits.odes import ode
import scikits.odes.sundials as sundials

from .deriv_funcs import (
    dderiv_B_φ_soln, taylor_series,
)

from ..file_format import LATEST_NAMESPACE as namespace

INTEGRATOR = "cvode"
COORDS = "spherical midplane 0"

log = logbook.Logger(__name__)


def ode_system(
    β, c_s, central_mass, η_O, η_A, η_H,
    taylor_stop_angle, init_con,
):
    """
    Set up the system we are solving for.
    """
    dderiv_ρ_M, dderiv_v_rM, dderiv_v_φM = taylor_series(
        β, c_s, η_O, η_A, η_H, init_con
    )

    internal_data = namespace.internal_data(  # pylint: disable=no-member
        derivs=[], params=[], angles=[], v_r_normal=[], v_φ_normal=[],
        ρ_normal=[], v_r_taylor=[], v_φ_taylor=[], ρ_taylor=[],
    )

    derivs_list = internal_data.derivs
    params_list = internal_data.params
    angles_list = internal_data.angles
    v_r_nlist = internal_data.v_r_normal
    v_φ_nlist = internal_data.v_φ_normal
    ρ_nlist = internal_data.ρ_normal
    v_r_taylist = internal_data.v_r_taylor
    v_φ_taylist = internal_data.v_φ_taylor
    ρ_taylist = internal_data.ρ_taylor

    def rhs_equation(θ, params, derivs):
        """
        Compute the ODEs
        """
        nonlocal derivs_list, params_list, angles_list
        nonlocal v_r_nlist, v_r_taylist
        nonlocal v_φ_nlist, v_φ_taylist
        nonlocal ρ_nlist, ρ_taylist

        params_list.append(np.copy(params))

        B_r = params[0]
        B_φ = params[1]
        B_θ = params[2]
        v_r = params[3]
        v_φ = params[4]
        v_θ = params[5]
        ρ = params[6]
        B_φ_prime = params[7]

        B_mag = sqrt(B_r**2 + B_φ**2 + B_θ**2)
        norm_B_r, norm_B_φ, norm_B_θ = (
            B_r/B_mag, B_φ/B_mag, B_θ/B_mag
        )

        deriv_B_φ = B_φ_prime
        deriv_B_r = (
            B_θ * (1 - β) +
            (
                v_θ * B_r - v_r * B_θ + deriv_B_φ * (
                    η_H * norm_B_θ -
                    η_A * norm_B_r * norm_B_φ
                ) + B_φ * (
                    η_H * (
                        norm_B_r * (1 - β) -
                        norm_B_θ * tan(θ)
                    ) - η_A * norm_B_φ * (
                        norm_B_θ * (1 - β) -
                        norm_B_r * tan(θ)
                    )
                )
            ) / (
                η_O + η_A * (1 - norm_B_φ**2)
            )
        )

        deriv_B_θ = (β - 2) * B_r + B_θ * tan(θ)

        deriv_v_r_taylor = θ * dderiv_v_rM
        deriv_v_r_normal = (
            v_r**2 / 2 +
            v_θ**2 + v_φ**2 +
            (2 * β * c_s**2) - central_mass +
            (
                B_θ * deriv_B_r + (β - 1)*(B_θ**2 + B_φ**2)
            ) / (
                4 * pi * ρ
            )
        ) / v_θ
        deriv_v_r = (
            deriv_v_r_taylor if θ < taylor_stop_angle else deriv_v_r_normal
        )
        if __debug__:
            if central_mass < (v_θ * deriv_v_r_normal + central_mass):
                log.info("not grav dominant at {}".format(degrees(θ)))

        deriv_v_φ_taylor = θ * dderiv_v_φM
        deriv_v_φ_normal = (
            (
                B_θ * B_φ_prime +
                (1-β) * B_r * B_φ +
                tan(θ) * B_θ * B_φ
            ) / (
                4 * pi * ρ
            ) - tan(θ) * v_φ * v_θ -
            (v_φ * v_r) / 2
        ) / v_θ
        deriv_v_φ = (
            deriv_v_φ_taylor if θ < taylor_stop_angle else deriv_v_φ_normal
        )

        deriv_v_θ = (
            v_r * (
                v_θ**2 / 2 + c_s**2 * (2 * β - 5/2)
            ) + v_θ * (
                tan(θ) * (v_φ**2 + c_s**2) + (
                    (β-1) * B_θ * B_r + B_r * deriv_B_r + B_φ * B_φ_prime -
                    B_φ**2 * tan(θ)
                ) / (4*pi * ρ)
            )
        ) / ((c_s - v_θ) * (c_s + v_θ))

        deriv_ρ_taylor = θ * dderiv_ρ_M
        deriv_ρ_normal = - ρ * (
            (
                (5/2 - 2 * β) * v_r + deriv_v_θ
            ) / v_θ - tan(θ)
        )
        deriv_ρ = deriv_ρ_taylor if θ < taylor_stop_angle else deriv_ρ_normal

        dderiv_B_φ = dderiv_B_φ_soln(
            B_r, B_φ, B_θ, η_O, η_H, η_A, θ, v_r,
            v_θ, v_φ, deriv_v_r, deriv_v_θ, deriv_v_φ,
            deriv_B_r, deriv_B_θ, deriv_B_φ, β
        )

        derivs[0] = deriv_B_r
        derivs[1] = deriv_B_φ
        derivs[2] = deriv_B_θ
        derivs[3] = deriv_v_r
        derivs[4] = deriv_v_φ
        derivs[5] = deriv_v_θ
        derivs[6] = deriv_ρ
        derivs[7] = dderiv_B_φ
        if __debug__:
            log.debug("θ: {}, {}", θ, degrees(θ))

        derivs_list.append(np.copy(derivs))
        angles_list.append(θ)
        v_r_nlist.append(deriv_v_r_normal)
        v_φ_nlist.append(deriv_v_φ_normal)
        ρ_nlist.append(deriv_ρ_normal)
        v_r_taylist.append(deriv_v_r_taylor)
        v_φ_taylist.append(deriv_v_φ_taylor)
        ρ_taylist.append(deriv_ρ_taylor)

        return 0
    return rhs_equation, internal_data


def solution(
        angles, initial_conditions, β, c_s, central_mass,
        η_O, η_A, η_H, relative_tolerance=1e-6,
        absolute_tolerance=1e-10,
        max_steps=500, taylor_stop_angle=0
):
    """
    Find solution
    """
    system, internal_data = ode_system(
        β, c_s, central_mass, η_O, η_A, η_H,
        radians(taylor_stop_angle), initial_conditions
    )
    solver = ode(
        INTEGRATOR, system,
        rtol=relative_tolerance,
        atol=absolute_tolerance,
        max_steps=max_steps,
        validate_flags=True,
        err_handler=ode_error_handler
    )
    try:
        soln = solver.solve(angles, initial_conditions)
    except sundials.CVODESolveFailed as e:
        soln = e.soln
        log.error(
            "Solver stopped at {} with flag {!s}.\n{}".format(
                degrees(soln.errors.t), soln.flag, soln.message
            )
        )
    return (
        soln.values.t, soln.values.y, internal_data,
        namespace.solution_properties(  # pylint: disable=no-member
            flag=soln.flag,
            coordinate_system=COORDS,
        )
    )


def ode_error_handler(error_code, module, func, msg, user_data):
    """ drop all CVODE messages """
    # pylint: disable=unused-argument
    pass
