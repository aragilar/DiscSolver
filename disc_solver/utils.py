# -*- coding: utf-8 -*-
"""
Useful functions
"""

from math import pi, cos, sin

import numpy as np
from matplotlib.ticker import FuncFormatter


def is_supersonic(v, B, rho, sound_speed, mhd_wave_type):
    """
    Checks whether velocity is supersonic.
    """
    v_axis = 1 if v.ndim > 1 else 0
    B_axis = 1 if B.ndim > 1 else 0

    v_sq = np.sum(v**2, axis=v_axis)
    B_sq = np.sum(B**2, axis=B_axis)

    cos_sq_psi = np.sum(
        (v.T**2/v_sq).T * (B.T**2/B_sq).T, axis=max(v_axis, B_axis)
    ) ** 2

    v_a_sq = B_sq / (4*pi*rho)
    if mhd_wave_type == "slow":
        mhd_wave_speed = 1/2 * (
            v_a_sq + sound_speed**2 - np.sqrt(
                (v_a_sq + sound_speed**2)**2 -
                4 * v_a_sq * sound_speed**2 * cos_sq_psi
            )
        )
    elif mhd_wave_type == "alfven":
        mhd_wave_speed = v_a_sq * cos_sq_psi
    elif mhd_wave_type == "fast":
        mhd_wave_speed = 1/2 * (
            v_a_sq + sound_speed**2 + np.sqrt(
                (v_a_sq + sound_speed**2)**2 -
                4 * v_a_sq * sound_speed**2 * cos_sq_psi
            )
        )

    return v_sq > mhd_wave_speed


def cot(angle):
    """
    Computes cot
    """
    if angle == pi:
        return float('-inf')
    elif angle == pi/2:
        return 0
    elif angle == 0:
        return float('inf')
    return cos(angle)/sin(angle)


def better_sci_format(physical_axis):
    """
    Use scientific notation for each tick mark for axis `physical_axis`.
    """
    physical_axis.set_major_formatter(
        FuncFormatter(lambda x, pos: "{:.2e}".format(x).replace("-", "âˆ’"))
    )
