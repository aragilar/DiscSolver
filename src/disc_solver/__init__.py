# -*- coding: utf-8 -*-
"""
Solver for PHD Project
"""
from ._version import version as __version__  # noqa
