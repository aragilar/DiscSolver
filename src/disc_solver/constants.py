# -*- coding: utf-8 -*-
"""
Physical constants used in code
"""
from .float_handling import float_type
G = float_type(6.6742e-8)  # (cm^3 âˆ• (g*s^2))
AU = float_type(1.4957871e13)  # cm
M_SUN = float_type(2e33)  # g
KM = float_type(1e5)  # cm
YEAR = float_type(31557600)  # s
