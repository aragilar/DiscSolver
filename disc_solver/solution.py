# -*- coding: utf-8 -*-
"""
The system of odes
"""

from math import pi, sqrt

import logbook

import numpy as np

from scikits.odes import ode
import scikits.odes.sundials as sundials

from .utils import cot
from .deriv_funcs import (
    dderiv_B_φ_soln, dderiv_v_φ_midplane, dderiv_v_r_midplane,
    dderiv_ρ_midplane,
)

INTEGRATOR = "cvode"

log = logbook.Logger(__name__)


def ode_system(
    β, c_s, central_mass, η_O, η_A, η_H,
    taylor_stop_angle, init_con,
):
    """
    Set up the system we are solving for.
    """
    B_θM = init_con[2]
    v_rM = init_con[3]
    v_φM = init_con[4]
    ρ_M = init_con[6]
    deriv_B_φM = init_con[7]

    deriv_v_θM = v_rM * (2 * β - 5/2)
    deriv_B_rM = - (
        B_θM * (1 - β) + (v_rM * B_θM + deriv_B_φM * η_H) / (η_O + η_A)
    )

    dderiv_B_θM = (β - 2) * deriv_B_rM + B_θM

    dderiv_ρ_M = dderiv_ρ_midplane(
        ρ_M, B_θM, v_rM, v_φM, deriv_B_rM, deriv_B_φM, β, c_s
    )

    dderiv_v_rM = dderiv_v_r_midplane(
        ρ_M, B_θM, v_rM, v_φM, deriv_v_θM, deriv_B_rM, deriv_B_φM, dderiv_B_θM,
        dderiv_ρ_M, β, η_O, η_H, η_A
    )

    dderiv_v_φM = dderiv_v_φ_midplane(
        ρ_M, B_θM, v_rM, v_φM, deriv_v_θM, deriv_B_rM, deriv_B_φM, dderiv_B_θM,
        dderiv_v_rM, dderiv_ρ_M, β, η_O, η_H, η_A
    )

    log.info("B_θ'': {}".format(dderiv_B_θM))
    log.info("v_r'': {}".format(dderiv_v_rM))
    log.info("v_φ'': {}".format(dderiv_v_φM))
    log.info("ρ'': {}".format(dderiv_ρ_M))

    internal_data = {
        "derivs": [],
        "params": [],
        "angles": [],
        "v_r normal": [],
        "v_φ normal": [],
        "ρ normal": [],
        "v_r taylor": [],
        "v_φ taylor": [],
        "ρ taylor": [],
    }
    derivs_list = internal_data["derivs"]
    params_list = internal_data["params"]
    angles_list = internal_data["angles"]
    v_r_nlist = internal_data["v_r normal"]
    v_φ_nlist = internal_data["v_φ normal"]
    ρ_nlist = internal_data["ρ normal"]
    v_r_taylist = internal_data["v_r taylor"]
    v_φ_taylist = internal_data["v_φ taylor"]
    ρ_taylist = internal_data["ρ taylor"]

    def rhs_equation(θ, params, derivs):
        """
        Compute the ODEs
        """
        nonlocal derivs_list, params_list, angles_list
        nonlocal v_r_nlist, v_r_taylist
        nonlocal v_φ_nlist, v_φ_taylist
        nonlocal ρ_nlist, ρ_taylist

        params_list.append(np.copy(params))

        B_r = params[0]
        B_φ = params[1]
        B_θ = params[2]
        v_r = params[3]
        v_φ = params[4]
        v_θ = params[5]
        ρ = params[6]
        B_φ_prime = params[7]

        B_mag = sqrt(B_r**2 + B_φ**2 + B_θ**2)
        norm_B_r, norm_B_φ, norm_B_θ = (
            B_r/B_mag, B_φ/B_mag, B_θ/B_mag
        )

        deriv_B_φ = B_φ_prime
        deriv_B_r = - (
            B_θ * (1 - β) +
            (
                v_r * B_θ - v_θ * B_r + deriv_B_φ * (
                    η_H * norm_B_θ -
                    η_A * norm_B_r * norm_B_φ
                ) + B_φ * (
                    η_H * (
                        norm_B_θ * cot(θ) -
                        norm_B_r * (1 - β)
                    ) + η_A * norm_B_φ * (
                        norm_B_θ * (1 - β) -
                        norm_B_r * cot(θ)
                    )
                )
            ) / (
                η_O + η_A * (1 - norm_B_φ**2)
            )
        )

        deriv_B_θ = (β - 2) * B_r - B_θ * cot(θ)

        deriv_v_r_taylor = (θ - pi/2) * dderiv_v_rM
        deriv_v_r_normal = (
            v_r**2 / 2 +
            v_θ**2 + v_φ**2 +
            (2 * β * c_s**2) - central_mass +
            (
                B_θ * deriv_B_r + (β - 1)*(B_θ**2 + B_φ**2)
            ) / (
                4 * pi * ρ
            )
        ) / v_θ
        deriv_v_r = (
            deriv_v_r_taylor if θ > taylor_stop_angle else deriv_v_r_normal
        )
        log.debug("v_r pos: {}", (
            v_r**2 / 2 + v_φ**2 + (2 * β * c_s**2) +
            (B_θ * deriv_B_r + (β - 1)*(B_θ**2 + B_φ**2)) / (4 * pi * ρ)
        ))
        log.debug("v_r neg: {}", - central_mass)

        deriv_v_φ_taylor = (θ - pi/2) * dderiv_v_φM
        deriv_v_φ_normal = (
            cot(θ) * v_φ * v_θ -
            (v_φ * v_r) / 2 +
            (
                B_θ * B_φ_prime +
                (1-β) * B_r * B_φ -
                cot(θ) * B_θ * B_φ
            ) / (
                4 * pi * ρ
            )
        ) / v_θ
        deriv_v_φ = (
            deriv_v_φ_taylor if θ > taylor_stop_angle else deriv_v_φ_normal
        )
        log.debug("v_φ A: {}", - (v_φ * v_r) / 2)
        log.debug("v_φ B: {}", B_θ * B_φ_prime / (4 * pi * ρ))
        log.debug("v_φ C: {}", (1-β) * B_r * B_φ / (4*pi*ρ))
        log.debug("v_φ D: {}", - cot(θ) * B_θ * B_φ / (4*pi*ρ))

        deriv_v_θ = (
            2 * (β-1) * v_r * c_s**2 / (c_s**2 - v_θ**2) - v_r / 2 -
            v_θ / (c_s**2 - v_θ**2) * (
                cot(θ) * v_φ**2 + c_s**2 * cot(θ) - (
                    (β-1) * B_θ * B_r + B_r * deriv_B_r + B_φ * B_φ_prime +
                    B_φ**2 * cot(θ)
                ) / (4*pi * ρ)
            )
        )

        deriv_ρ_taylor = (θ - pi/2) * dderiv_ρ_M
        deriv_ρ_normal = - ρ * (
            (
                (5/2 - 2 * β) * v_r + deriv_v_θ
            ) / v_θ + cot(θ)
        )
        deriv_ρ = deriv_ρ_taylor if θ > taylor_stop_angle else deriv_ρ_normal

        dderiv_B_φ = dderiv_B_φ_soln(
            B_r, B_φ, B_θ, η_O, η_H, η_A, θ, v_r,
            v_θ, v_φ, deriv_v_r, deriv_v_θ, deriv_v_φ,
            deriv_B_r, deriv_B_θ, deriv_B_φ, β
        )

        derivs[0] = deriv_B_r
        derivs[1] = deriv_B_φ
        derivs[2] = deriv_B_θ
        derivs[3] = deriv_v_r
        derivs[4] = deriv_v_φ
        derivs[5] = deriv_v_θ
        derivs[6] = deriv_ρ
        derivs[7] = dderiv_B_φ

        log.debug("θ: {}, {}", θ, θ/pi*180)

        derivs_list.append(np.copy(derivs))
        angles_list.append(θ)
        v_r_nlist.append(deriv_v_r_normal)
        v_φ_nlist.append(deriv_v_φ_normal)
        ρ_nlist.append(deriv_ρ_normal)
        v_r_taylist.append(deriv_v_r_taylor)
        v_φ_taylist.append(deriv_v_φ_taylor)
        ρ_taylist.append(deriv_ρ_taylor)

        return 0
    return rhs_equation, internal_data


def solution(
        angles, initial_conditions, β, c_s, central_mass,
        η_O, η_A, η_H, relative_tolerance=1e-6,
        absolute_tolerance=1e-10,
        max_steps=500, taylor_stop_angle=pi/2
):
    """
    Find solution
    """
    system, internal_data = ode_system(
        β, c_s, central_mass, η_O, η_A, η_H,
        taylor_stop_angle / 180 * pi, initial_conditions
    )
    solver = ode(
        INTEGRATOR, system,
        rtol=relative_tolerance,
        atol=absolute_tolerance,
        max_steps=max_steps,
        validate_flags=True,
        err_handler=ode_error_handler
    )
    try:
        soln = solver.solve(angles, initial_conditions)
    except sundials.CVODESolveFailed as e:
        log.error(e)
        soln = e.soln
    return soln.values.t, soln.values.y, internal_data, soln.flag


def ode_error_handler(error_code, module, func, msg, user_data):
    """ drop all CVODE messages """
    # pylint: disable=unused-argument
    pass
