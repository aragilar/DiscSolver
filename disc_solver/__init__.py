# -*- coding: utf-8 -*-
"""
Solver for PHD Project
"""

__version__ = "0.1"

import argparse
from pprint import pprint
import signal
import tempfile
from types import SimpleNamespace

import arrow
import logbook
from logbook.compat import redirected_warnings, redirected_logging

import numpy as np
import h5py

import matplotlib as mpl
mpl.use("Qt4Agg")
mpl.rcParams["backend.qt4"] = "PySide"
import matplotlib.pyplot as plt

from .analyse import generate_plot, info
from .config import define_conditions, get_input
from .solution import solution

log = logbook.Logger(__name__)


def stack_info(frame):
    """Walk up stack printing line and locals"""
    print('f_code', frame.f_code)
    print('f_locals:')
    pprint(frame.f_locals)
    print()
    if frame.f_back:
        stack_info(frame.f_back)


def sig_handler(signum, frame):
    """Custom signal handler"""
    print(signum)
    stack_info(frame)


def solution_main(output_file=None, get_config_file=True):
    """
    Main function to generate solution
    """
    inp, = get_input(get_file=get_config_file)
    if output_file:
        inp.output_file = output_file
    cons = define_conditions(inp)

    signal.signal(signal.SIGINT, sig_handler)

    null_handler = logbook.NullHandler()
    with null_handler.applicationbound():
        angles, soln = solution(
            cons.angles, cons.init_con, inp.β, cons.c_s, cons.norm_kepler_sq,
            cons.η_O, cons.η_A, cons.η_H, max_steps=inp.max_steps,
            taylor_stop_angle=inp.taylor_stop_angle
        )

    with h5py.File(inp.output_file) as f:
        grp = f.create_group(str(arrow.now()))
        grp['angles'] = angles
        grp['solution'] = soln
        grp.attrs.update(vars(inp))


def analyse_main(output_file=None, **kwargs):
    """
    Main function to analyse solution
    """
    if output_file is None:
        output_file, kwargs = analyse_parser(kwargs)
    make_plot = kwargs.get("show") or kwargs.get("output_fig_file")

    with redirected_warnings(), redirected_logging():
        with h5py.File(output_file) as f:
            for grp in f.values():
                inp = SimpleNamespace(**grp.attrs)  # pylint: disable=star-args
                cons = define_conditions(inp)
                angles = np.array(grp["angles"])
                soln = np.array(grp["solution"])
                if make_plot:
                    fig = generate_plot(angles, soln, inp, cons)
                if kwargs.get("show"):
                    plt.show()
                if kwargs.get("output_fig_file"):
                    fig.savefig(kwargs.get("output_fig_file"))
                if kwargs.get("info"):
                    info(inp, cons)


def analyse_parser(kwargs):
    """
    CLI Parser for analyse
    """
    parser = argparse.ArgumentParser(description='Analyser for DiscSolver')
    parser.add_argument("output_file")

    if not kwargs:
        parser.add_argument(
            "--show", "-s", action="store_true", default=False
        )
        parser.add_argument("--output_fig_file", "-o")
        parser.add_argument(
            "--info", "-i", action="store_true", default=False
        )

    output_file = parser.parse_args().output_file

    if not kwargs:
        kwargs = vars(parser.parse_args())
    return output_file, kwargs


def main():
    """
    The main function
    """
    null_handler = logbook.NullHandler()
    with null_handler.applicationbound():
        with tempfile.NamedTemporaryFile() as output_file:
            solution_main(output_file=output_file.name, get_config_file=False)
            analyse_main(
                output_file=output_file.name, show=True,
                output_fig_file="plot.png"
            )
