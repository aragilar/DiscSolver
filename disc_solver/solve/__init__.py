# -*- coding: utf-8 -*-
"""
Solver component of Disc Solver
"""
import argparse

import arrow
import logbook
from logbook.compat import redirected_warnings, redirected_logging

from h5py import File

from .config import get_input, step_input
from .solution import create_soln_splitter, solver_generator
from .stepper import (
    binary_searcher, stepper_creator, writer_generator, cleanup_generator
)

from ..file_format import wrap_hdf5_file
from ..logging import logging_options, log_handler
from ..utils import allvars as vars

log = logbook.Logger(__name__)


def solution_main(output_file=None, ismain=True):
    """
    Main function to generate solution
    """
    if ismain:
        parser = argparse.ArgumentParser(description='Solver for DiscSolver')
        parser.add_argument("conffile")
        logging_options(parser)
        args = vars(parser.parse_args())
        conffile = args["conffile"]
    else:
        args = {
            "quiet": True,
        }
        conffile = None
    gen_file_name = True if output_file is None else False
    with log_handler(args), redirected_warnings(), redirected_logging():
        config_inp = get_input(conffile=conffile)
        step_func = step_input(config_inp)
        if gen_file_name:
            output_file = config_inp.label + str(arrow.now()) + ".hdf5"
        with File(output_file) as f:
            soln_file = wrap_hdf5_file(f, version="latest", new=True)
            soln_file.root.config_input = config_inp
            soln_file.root.config_filename = str(conffile)
            writer = writer_generator(soln_file)
            cleanup = cleanup_generator(soln_file, writer)
            binary_searcher(
                solver_generator(), cleanup,
                stepper_creator(
                    writer, step_func,
                    create_soln_splitter("v_Î¸_deriv")
                ),
                config_inp,
            )
