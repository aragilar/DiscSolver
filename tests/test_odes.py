
# -*- coding: utf-8 -*-

from math import pi, sqrt, tan
import unittest

import logbook

import numpy as np

from disc_solver.constants import G
from disc_solver.utils import ODEIndex
from disc_solver.solve.solution import ode_system

ODE_NUMBER = 11

class ODE_Test(unittest.TestCase):
    def setUp(self):
        # These are all arbitrary values, this should not affect the result
        params = np.array([7 for i in range(ODE_NUMBER)])
        self.β = 2
        self.norm_kepler_sq = 2
        self.a_0 = 2

        # This is slightly off the plane, this should mean we don't get
        # cancellation
        self.angle = 0.1

        γ = 5/4 - self.β

        # Here's the action computation
        derivs = np.zeros(ODE_NUMBER)
        with logbook.NullHandler().applicationbound():
            self.rhs, internal_data = ode_system(
                γ=γ, a_0=self.a_0, norm_kepler_sq=self.norm_kepler_sq,
                init_con=params, with_taylor=False
            )
            self.rhs(self.angle, params, derivs)

        # Assign useful names to params, derivs
        self.B_r = params[ODEIndex.B_r]
        self.B_φ = params[ODEIndex.B_φ]
        self.B_θ = params[ODEIndex.B_θ]
        self.v_r = params[ODEIndex.v_r]
        self.v_φ = params[ODEIndex.v_φ]
        self.v_θ = params[ODEIndex.v_θ]
        self.ρ = params[ODEIndex.ρ]
        self.η_O = params[ODEIndex.η_O]
        self.η_A = params[ODEIndex.η_A]
        self.η_H = params[ODEIndex.η_H]
        self.deriv_B_r = derivs[ODEIndex.B_r]
        self.deriv_B_φ = derivs[ODEIndex.B_φ]
        self.deriv_B_θ = derivs[ODEIndex.B_θ]
        self.deriv_v_r = derivs[ODEIndex.v_r]
        self.deriv_v_φ = derivs[ODEIndex.v_φ]
        self.deriv_v_θ = derivs[ODEIndex.v_θ]
        self.deriv_ρ = derivs[ODEIndex.ρ]
        self.dderiv_B_φ = derivs[ODEIndex.B_φ_prime]
        self.deriv_η_O = derivs[ODEIndex.η_O]
        self.deriv_η_A = derivs[ODEIndex.η_A]
        self.deriv_η_H = derivs[ODEIndex.η_H]


        B_mag = sqrt(self.B_r**2 + self.B_φ**2 + self.B_θ **2)
        self.norm_B_r, self.norm_B_φ, self.norm_B_θ = (
            self.B_r/B_mag, self.B_φ/B_mag, self.B_θ/B_mag)

        self.params = params
        self.derivs = derivs

    def test_continuity(self):
        eqn = (
            (5/2 - 2 * self.β) * self.v_r +
            self.deriv_v_θ + (
                self.v_θ / self.ρ
            ) * (
                self.deriv_ρ - self.ρ * tan(self.angle)
            )
        )
        print("continuity", eqn)
        self.assertAlmostEqual(0, eqn)

    def test_solenoid(self):
        eqn = self.deriv_B_θ - (
            (self.β - 2) * self.B_r + self.B_θ * tan(self.angle)
        )
        print("solenoid", eqn)
        self.assertAlmostEqual(0, eqn)

    def test_radial_momentum(self):
        eqn = (self.v_θ * self.deriv_v_r - 1/2 * self.v_r**2 -
            self.v_θ**2 - self.v_φ**2 + self.norm_kepler_sq -
            2 * self.β - self.a_0 / self.ρ * (
                self.B_θ * self.deriv_B_r + (self.β - 1) * (
                    self.B_θ**2 + self.B_φ**2
                )
            )
        )
        print("radial momentum", eqn)
        self.assertAlmostEqual(0, eqn)

    def test_azimuthal_mometum(self):
        eqn = (self.v_θ * self.deriv_v_φ + 1/2 * self.v_r * self.v_φ -
            tan(self.angle) * self.v_θ * self.v_φ - self.a_0 / self.ρ * (
                self.B_θ * self.deriv_B_φ +
                (1 - self.β) * self.B_r * self.B_φ -
                tan(self.angle) * self.B_θ * self.B_φ
            )
        )
        print("azimuthal momentum", eqn)
        self.assertAlmostEqual(0, eqn)

    def test_polar_momentum(self):
        eqn = (self.v_r * self.v_θ / 2 + self.v_θ * self.deriv_v_θ +
            tan(self.angle) * self.v_φ ** 2 + self.deriv_ρ / self.ρ +
            self.a_0 / self.ρ * (
                (self.β - 1) * self.B_θ * self.B_r +
                self.B_r * self.deriv_B_r + self.B_φ * self.deriv_B_φ -
                self.B_φ ** 2 * tan(self.angle)
            )
        )
        print("polar momentum", eqn)
        self.assertAlmostEqual(0, eqn)

    def test_polar_induction(self):
        eqn = (
            self.v_θ * self.B_r -
            self.v_r * self.B_θ + (
                self.B_θ * (1 - self.β) - self.deriv_B_r
            ) * (
                self.η_O + self.η_A * (1 - self.norm_B_φ**2)
            ) + self.deriv_B_φ * (
                self.η_H * self.norm_B_θ -
                self.η_A * self.norm_B_r * self.norm_B_φ
            ) + self.B_φ * (
                self.η_H * (
                    self.norm_B_r * (1 - self.β) -
                    self.norm_B_θ * tan(self.angle)
                ) - self.η_A * self.norm_B_φ * (
                    self.norm_B_θ * (1 - self.β) -
                    self.norm_B_r * tan(self.angle)
                )
            )
        )
        print("polar induction", eqn)
        self.assertAlmostEqual(0, eqn)

    def test_azimuthal_induction_numeric(self):
        step = 1e-4
        new_params = self.params + self.derivs * step
        new_angle = self.angle + step
        new_derivs = np.zeros(ODE_NUMBER)
        self.rhs(new_angle, new_params, new_derivs)
        dderiv_B_φ_hacked = ((new_derivs - self.derivs) / step)[1]
        eqn = dderiv_B_φ_hacked - self.dderiv_B_φ
        print("azimuthal induction (num)", eqn)
        self.assertAlmostEqual(0, eqn)

    # This would be useful to do when I have time
    #def test_azimuthal_induction_algebraic(self):
    #    eqn = 1 # FAIL
    #    print(eqn)
    #    self.assertAlmostEqual(0,eqn)

