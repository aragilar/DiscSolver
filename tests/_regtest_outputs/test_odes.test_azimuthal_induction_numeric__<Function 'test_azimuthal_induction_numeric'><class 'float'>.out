-1.630917623174355e-12
